From 6fa22981bae9ba78c042a6ef1d8073e224b5bb14 Mon Sep 17 00:00:00 2001
From: Alexandre Frade <kernel@xanmod.org>
Date: Fri, 20 Jun 2025 01:14:28 +0000
Subject: [PATCH 03/18] XANMOD: kbuild: Add SMS-based software pipelining flags

Add support for Swing Modulo Scheduling (SMS) when default optimization level
is enabled. This optimization performs software pipelining on innermost loops,
overlapping iterations to improve instruction-level parallelism.

Signed-off-by: Alexandre Frade <kernel@xanmod.org>
---
 Makefile | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/Makefile b/Makefile
index 43b112df616a..d48a189221da 100644
--- a/Makefile
+++ b/Makefile
@@ -939,9 +939,13 @@ KBUILD_CFLAGS += $(call cc-option,-mllvm=-polly -mllvm=-polly-ast-use-context \
 ifdef CONFIG_LD_DEAD_CODE_DATA_ELIMINATION
 KBUILD_CFLAGS += $(call cc-option,-mllvm=-polly-run-dce)
 endif
+# Enable software pipelining (Swing Modulo Scheduling) for innermost loops.
+KBUILD_CFLAGS += $(call cc-option,-fmodulo-sched -fmodulo-sched-allow-regmoves) \
+		 $(call cc-option,-mllvm=-enable-pipeliner)
 KBUILD_RUSTFLAGS += -Copt-level=2
 else ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
 KBUILD_CFLAGS += -Os
+KBUILD_CFLAGS += $(call cc-option,-mllvm=-enable-pipeliner-opt-size)
 KBUILD_RUSTFLAGS += -Copt-level=s
 endif
 
-- 
2.47.3

